File: js/src/history.ts
/** Bounded message log + per-sender resync watermark for GTP. */

/** One entry in {@link MessageHistory}. */
export interface MessageEntry {
    senderId: number;
    messageId: bigint;
    text: string;
}

/**
 * Bounded ring buffer of recent GTP messages.
 *
 * Used to serve resync requests from re-joining peers — keep a few
 * thousand recent messages, then `since(watermark)` returns everything
 * above the caller's high-water mark.
 */
export class MessageHistory {
    private readonly capacity: number;
    private buffer: MessageEntry[] = [];

    constructor(capacity: number) {
        if (capacity <= 0) throw new Error("capacity must be > 0");
        this.capacity = capacity;
    }

    get length(): number { return this.buffer.length; }

    /** Record a message. Returns `true` if newly added. */
    push(entry: MessageEntry): boolean {
        if (this.contains(entry.senderId, entry.messageId)) return false;
        if (this.buffer.length === this.capacity) this.buffer.shift();
        this.buffer.push(entry);
        return true;
    }

    /** `true` if `(senderId, messageId)` is present. */
    contains(senderId: number, messageId: bigint): boolean {
        return this.buffer.some(e => e.senderId === senderId && e.messageId === messageId);
    }

    /** Yields every message produced after the given watermark. */
    *since(watermark: Watermark): Iterable<MessageEntry> {
        for (const m of this.buffer) {
            const hw = watermark.lastSeen(m.senderId);
            if (hw === null || m.messageId > hw) yield m;
        }
    }

    /** Yields messages from a single sender newer than `sinceMessageId`. */
    *sinceForSender(senderId: number, sinceMessageId: bigint): Iterable<MessageEntry> {
        for (const m of this.buffer)
            if (m.senderId === senderId && m.messageId > sinceMessageId) yield m;
    }

    /** Drops every message in the buffer. */
    clear(): void { this.buffer = []; }
}

/** Per-sender high-water mark of accepted GTP `message_id`s. */
export class Watermark {
    private readonly _lastSeen = new Map<number, bigint>();

    /** Record that `messageId` from `senderId` has been observed. */
    observe(senderId: number, messageId: bigint): void {
        const prev = this._lastSeen.get(senderId) ?? 0n;
        if (messageId > prev) this._lastSeen.set(senderId, messageId);
    }

    /** Last observed `messageId` for `senderId`, or `null`. */
    lastSeen(senderId: number): bigint | null {
        return this._lastSeen.get(senderId) ?? null;
    }

    /** Returns a copy of the underlying map. */
    snapshot(): Map<number, bigint> { return new Map(this._lastSeen); }

    get size(): number { return this._lastSeen.size; }

    /** Drops every entry. */
    clear(): void { this._lastSeen.clear(); }
}

File: python/gbp_stack/history.py
"""Bounded message log + per-sender resync watermark for GTP."""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from typing import Dict, Iterator, Optional


@dataclass(frozen=True)
class MessageEntry:
    """One entry in :class:`MessageHistory`."""

    sender_id: int
    message_id: int
    text: str


class MessageHistory:
    """Bounded ring buffer of recent GTP messages.

    Used to serve resync requests from re-joining peers — keep a few
    thousand recent messages, then `since(watermark)` returns everything
    above the caller's high-water mark.
    """

    __slots__ = ("_capacity", "_buffer")

    def __init__(self, capacity: int) -> None:
        if capacity <= 0:
            raise ValueError("capacity must be > 0")
        self._capacity = capacity
        self._buffer: "deque[MessageEntry]" = deque(maxlen=capacity)

    def __len__(self) -> int:
        return len(self._buffer)

    def push(self, entry: MessageEntry) -> bool:
        """Record a message. Returns ``True`` if newly added."""
        if self.contains(entry.sender_id, entry.message_id):
            return False
        self._buffer.append(entry)
        return True

    def contains(self, sender_id: int, message_id: int) -> bool:
        """``True`` if ``(sender_id, message_id)`` is buffered."""
        return any(
            m.sender_id == sender_id and m.message_id == message_id
            for m in self._buffer
        )

    def since(self, watermark: "Watermark") -> Iterator[MessageEntry]:
        """Yields every message produced after ``watermark`` in insertion order."""
        for m in self._buffer:
            hw = watermark.last_seen(m.sender_id)
            if hw is None or m.message_id > hw:
                yield m

    def since_for_sender(self, sender_id: int, since_message_id: int) -> Iterator[MessageEntry]:
        """Yields messages from a single sender newer than ``since_message_id``."""
        for m in self._buffer:
            if m.sender_id == sender_id and m.message_id > since_message_id:
                yield m

    def clear(self) -> None:
        """Drops every message in the buffer."""
        self._buffer.clear()


class Watermark:
    """Per-sender high-water mark of accepted GTP ``message_id`` s."""

    __slots__ = ("_last_seen",)

    def __init__(self) -> None:
        self._last_seen: Dict[int, int] = {}

    def observe(self, sender_id: int, message_id: int) -> None:
        """Record that ``message_id`` from ``sender_id`` has been observed."""
        prev = self._last_seen.get(sender_id, 0)
        if message_id > prev:
            self._last_seen[sender_id] = message_id

    def last_seen(self, sender_id: int) -> Optional[int]:
        """Last observed ``message_id`` for ``sender_id`` (``None`` if unseen)."""
        return self._last_seen.get(sender_id)

    def snapshot(self) -> Dict[int, int]:
        """Returns a copy of the underlying mapping."""
        return dict(self._last_seen)

    def __len__(self) -> int:
        return len(self._last_seen)

    def clear(self) -> None:
        """Drops every entry."""
        self._last_seen.clear()
